machineconfiguration/v1/types.go

```diff
@@ -745,3 +749,169 @@ type ContainerRuntimeConfigList struct {
+type MachineState struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec MachineStateSpec `json:"spec"`
+	// +optional
+	Status MachineStateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineStateList describes all of the MachinesStates on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineStateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineState `json:"items"`
+}
+
+// MachineStateSpec describes the type of State we are managing
+type MachineStateSpec struct {
+	// Kind describes the type of MachineState
+	// +optional
+	Kind   MachineStateType   `json:"machineStateKind,omitempty"`
+	Config MachineStateConfig `json:"configuration"`
+}
+
+// MachineStateStatus holds the reported information on a particular MachineState
+type MachineStateStatus struct {
+	// Config Describes the Kind of MachineState
+	Config MachineStateConfig `json:"configuration"`
+	// MostRecentState MCO Object Mapped to a Progress Condition
+	MostRecentState map[string]ProgressionCondition `json:"progress"`
+	// Progression describes the 20 Most Recent MachineStates
+	Progression []ProgressionCondition `json:"allStates"`
+	// MostRecentError is populated if the State reports an error.
+	MostRecentError string `json:"error"`
+	// Healthy reports if the state is Healthy Overall
+	Healthy bool `json:"healthy"`
+}
+
+// MachineStateConfig describes the configuration of a MachineState
+type MachineStateConfig struct {
+	corev1.ObjectReference `json:",inline"`
+}
+
+// An OperatorObject is used as the key in the map describing the most recent states
+// for example, if the MachineStateKind == DaemonState, the Object will be a Node and we will report
+// each node's most recent state.
+type OperatorObject string
+
+const (
+	// MCP describes a MachineConfigPool
+	MCP OperatorObject = "MachineConfigPool"
+	// KC describes a KubeletConfig object
+	KC OperatorObject = "KubeletConfig"
+	// MC describes a MachineConfig object
+	MC OperatorObject = "MachineConfig"
+	// CC describes a ControllerConfig object
+	CC OperatorObject = "ControllerConfig"
+	// Node describes a Node object
+	Node OperatorObject = "Node"
+)
+
+// StateProgress is each possible state for each possible MachineStateType
+// UpgradeProgression Kind will only use the "MachinConfigPoolUpdate..." types for example
+type StateProgress string
+
+const (
+	// OperatorSyncRenderConfig describes a machine that is creating or syncing its render config
+	OperatorSyncRenderConfig StateProgress = "OperatorSyncRenderConfig"
+	// OperatorSyncMCP describes a machine that is syncing or applying its MachineConigPools
+	OperatorSyncMCP StateProgress = "OperatorSyncMCP"
```

Comment:
nit: I can mentally expand MCP -> MachineConfigPool, but I don't see a need to ask new customers to do that.  Can we expand the abbreviations in these state strings to make them more accessible?

Also, are these types we are committing to supporting as a stable API?  Or to we want this to be a "we'll try to pick strings you'll recognize" API, without committing to continuing to use the same strings for similar states in future controller releases?